Doc 7238 documentation for magma (#2309)
couchbase/docs-server: modules/learn/pages/buckets-memory-and-storage/backend-storage.adoc

= Backend Storage

> :page-pagination: 

Reviewer: I don't think we should have Next and Previous page links in the Learn section, it should be more of an "every page is page one" environment. Also, it'll be inconsistent if this subsection has Next and Previous links, and none of the other subsections in the Learn section have them. However, you should definitely include links to related topics at the end of the page, if required.

Author: Yes, you're right. A few other places in the docs use a capital S, but the design docs all use lower case. I'll correct the other mentions while I'm here.

> :description: pass:q[Couchbase supports two different backend storage mechanisms: `CouchStore` and `Magma`. It is important to understand which backend storage is best suited to your requirements.]

Reviewer: Is the CamelCase correct? All of the internal documentation I've seen capitalizes it as Couchstore.

```suggestion
:description: pass:q[Couchbase supports two different backend storage mechanisms: `Couchstore` and `Magma`. It is important to understand which backend storage is best suited to your requirements.]
```

Reviewer: Update throughout if necessary.

> :description: pass:q[Couchbase supports two different backend storage mechanisms: `CouchStore` and `Magma`. It is important to understand which backend storage is best suited to your requirements.]

Reviewer: I'm not sure about having formatting tags in the metadata and prefer to leave it as plain text (until someone can show me that it doesn't matter!) Definitely more readable without

```suggestion
:description: Couchbase supports two different backend storage mechanisms: Couchstore and Magma. It is important to understand which backend storage is best suited to your requirements.
```

Author: Yes, fair comment, since it ends up in the meta data, probably not a good place to format stuff.

> :page-pagination: prev

Reviewer: No pagination in Learn section (see comment above).

[abstract]
{description}

== CouchStore

> CouchStore is the default backend storage engine that has been in use for more than ten years. It's optimised for high performance with large data sets while using fewer system resources (the minimum bucket size for CouchStore backend is 100 MB). If you have a small data set that can fit in memory, then you should consider using CouchStore.

Reviewer: [Ventilated prose](https://docs.couchbase.com/home/contribute/pages.html#ventilate)

```suggestion
CouchStore is the default backend storage engine that has been in use for more than ten years.
It's optimised for high performance with large data sets while using fewer system resources (the minimum bucket size for CouchStore backend is 100 MB).
If you have a small data set that can fit in memory, then you should consider using CouchStore.
```

== Magma

> Magma is Couchbase's latest backend engine is designed for high performance with larger datasets (the smallest bucket you can have with a Magma backend is 256 MB). It is optimised for SSD disk access and higher transaction performance, and has such has higher hardware resource requirements. It is especially suited to data sets that will not fit into available memory.

Reviewer: [Ventilated prose](https://docs.couchbase.com/home/contribute/pages.html#ventilate)

```suggestion
Magma is Couchbase's latest backend engine is designed for high performance with larger datasets (the smallest bucket you can have with a Magma backend is 256 MB).
It is optimised for SSD disk access and higher transaction performance, and has such has higher hardware resource requirements.
It is especially suited to data sets that will not fit into available memory.
```

= Backend Storage
:description: pass:q[Couchbase supports two different backend storage mechanisms: Couchstore and Magma. It is important to understand which backend storage is best suited to your requirements.]

[abstract]
{description}

== Couchstore

> Couchstore is the default backend storage engine that has been in use for more than ten years. 

Reviewer 2: replace 'backend' with 'bucket'

> It's optimised for high performance with large data sets while using fewer system resources (the minimum bucket size for Couchstore backend is 100 MB). 

Reviewer 2: ```suggestion
It's the ideal choice for high performance caching use cases with high working sets (> 20%). Couchstore requires lower minimum resources (the minimum bucket size for Couchstore backend is 100 MB) and is a good choice for use cases on low-end hardware. 
```

> If you have a small data set that can fit in memory, then you should consider using Couchstore.

Reviewer 2: not needed. incorporated in above comment.

== Magma

> Magma is Couchbase's latest backend engine is designed for high performance with larger datasets (the smallest bucket you can have with a Magma backend is 256 MB). 

Reviewer 2: ```suggestion
Magma is Couchbase's latest storage engine that is designed for high performance with very large datasets that do not fit in memory. It is ideal for such disk-oriented use cases that primarily rely on disk access. The performance of disk access will be as good as the underlying disk subsystems. e.g. using NVMe SSDs will give higher performance.  
Magma can work with very low amounts of memory for large datasets. e.g. for a node holding 5TB of data, Magma can be used with only 64GB RAM. Magma buckets do require a minimum of 256MB RAM and you should not go below a RAM allocation of 1% of your total data size. 
```

> It is optimised for SSD disk access and higher transaction performance, and has such has higher hardware resource requirements. 

Reviewer 2: not needed. see above comment.

> It is especially suited to data sets that will not fit into available memory.

Reviewer 2: not needed.

> IMPORTANT: Currently, Magma does not support Full Text Searching, Eventing or Analytics.

Reviewer 2: ```suggestion
IMPORTANT: Currently, Magma does not support Full Text Searching, Eventing or Analytics. If the use case requires these features, Couchstore should be used.
```

== Couchstore and Magma at a Glance

|===
| {empty} | Couchstore |Magma

| Minimum bucket size
| 100 MB
| 256 MB

> | Maximum storage per node

Reviewer 2: ```suggestion
| Maximum data per node
```

| 3 TB
| 5 TB

> | Maximum storage per cluster

Reviewer 2: remove this one.  We don't really mandate a limit of 10 nodes per cluster.

| 30 TB
| 100+ TB

| Full text search
| Yes
| No

| Eventing
| Yes
| No

| Analytics
| Yes
| No
|===

> == When should you use Couchstore?

Reviewer 2: Add the following sentence.
The choice of Couchstore or Magma is available at the Bucket level. A single Couchbase cluster can have a mix of Couchstore and Magma buckets.

You should use the Couchstore backend if:

> * you have a dataset that will fit in available memory.

Reviewer 2: ```suggestion
* you have a dataset whose working set will fit in available memory (and the working set is > 20%)
```

* you're running the Couchbase server on low-end hardware.

> * your Couchbase server needs to support full text searching, eventing, or analytics.

Reviewer 2: ```suggestion
* your bucket  needs to support full text searching, eventing, or analytics.
```

== When should you use Magma?

You should use the Magma backend if:

> * your data set won't fit in available memory.

Reviewer 2: ```suggestion
* your data set is much larger than the available memory (available memory < 20% of your dataset) and you need disk access speed only
```

> * you need to store large amounts of data in a single node (> 5 TB)

Reviewer 2: ```suggestion
* you need to store large amounts of data in a single node (> 5 TB) using lowest amount of memory
```

> * your application(s) makes heavy use of transactions.

Reviewer 2: ```suggestion
* your application(s) makes heavy use of transactions with persistence based durability.
```
